server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -470,7 +514,104 @@ protected long currentNanoTime() {
     }
 
     public void cleanCaches() {
-        gatewayAllocator.cleanCaches();
+        assert assertInitialized();
+        existingShardsAllocators.values().forEach(ExistingShardsAllocator::cleanCaches);
+    }
+
+    public int getNumberOfInFlightFetches() {
+        assert assertInitialized();
+        return existingShardsAllocators.values().stream().mapToInt(ExistingShardsAllocator::getNumberOfInFlightFetches).sum();
+    }
+
+    public ShardAllocationDecision explainShardAllocation(ShardRouting shardRouting, RoutingAllocation allocation) {
+        assert allocation.debugDecision();
+        AllocateUnassignedDecision allocateDecision
+            = shardRouting.unassigned() ? explainUnassignedShardAllocation(shardRouting, allocation) : AllocateUnassignedDecision.NOT_TAKEN;
+        if (allocateDecision.isDecisionTaken()) {
+            return new ShardAllocationDecision(allocateDecision, MoveDecision.NOT_TAKEN);
+        } else {
+            return shardsAllocator.decideShardAllocation(shardRouting, allocation);
+        }
+    }
+
+    private AllocateUnassignedDecision explainUnassignedShardAllocation(ShardRouting shardRouting, RoutingAllocation routingAllocation) {
+        assert shardRouting.unassigned();
+        assert routingAllocation.debugDecision();
+        assert assertInitialized();
+        final ExistingShardsAllocator existingShardsAllocator = getAllocatorForShard(shardRouting, routingAllocation);
+        final AllocateUnassignedDecision decision
+            = existingShardsAllocator.explainUnassignedShardAllocation(shardRouting, routingAllocation);
+        if (decision.isDecisionTaken()) {
+            return decision;
+        }
+        return AllocateUnassignedDecision.NOT_TAKEN;
+    }
+
+    private ExistingShardsAllocator getAllocatorForShard(ShardRouting shardRouting, RoutingAllocation routingAllocation) {
+        assert assertInitialized();
+        final String allocatorName = ExistingShardsAllocator.EXISTING_SHARDS_ALLOCATOR_SETTING.get(
+            routingAllocation.metaData().getIndexSafe(shardRouting.index()).getSettings());
```

Comment:
It involves passing the `IndexMetaData` around to a lot of places. I mean we _could_ do that if you're sure this is sufficiently expensive, but it seems surprising to me  that this would be a concern.